src/node/internal/internal_tls_wrap.ts

```diff
@@ -0,0 +1,714 @@
+    // When handshake fails control is not yet released,
+    // so self._tlsError will return null instead of actual error
+
+    // Set closing the socket after emitting an event since the socket needs to
+    // be accessible when the `tlsClientError` event is emitted.
+    this.destroy(err);
+  } else {
+    // Emit error
+    this._emitTLSError(err);
+  }
+}
+
+// Used by both client and server TLSSockets to start data flowing from _handle,
+// read(0) causes a StreamBase::ReadStart, via Socket._read.
+function initRead(tlsSocket: TLSSocket, socket: Socket | null): void {
+  // If we were destroyed already don't bother reading
+  if (!tlsSocket._handle) return;
+
+  // Socket already has some buffered data - emulate receiving it
+  if (socket?.readableLength) {
+    // @ts-expect-error TS6133 Disable unused variable warning
+    let buf: string | Buffer | null;
+    // Eslint is disabled due to @types/node missing types for
+    // Duplex.prototype.read(). It returns any...
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unused-vars
+    while ((buf = socket.read()) !== null) {
+      // This is one of the places where we differ from Node.js implementation
+      // Node.js implementation calls:
+      // tlsSocket._handle.receive(buf);
+    }
+  }
+
+  tlsSocket.read(0);
+}
+
+// @ts-expect-error TS2323 Cannot redeclare error.
+export function TLSSocket(
+  this: TLSSocket,
+  socket: Socket | Duplex | undefined,
+  opts: TLSSocket['_tlsOptions']
+): void {
+  const tlsOptions = { ...opts };
+
+  if (tlsOptions.enableTrace) {
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.enableTrace');
+  }
+
+  if (tlsOptions.isServer) {
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.isServer');
+  }
+
+  if (tlsOptions.server) {
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.server');
+  }
+
+  if (tlsOptions.requestCert) {
+    // Servers will request certificate from clients.
+    // Does not apply to Cloudflare Workers.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.requestCert');
+  }
+
+  if (tlsOptions.rejectUnauthorized) {
+    // Used by servers to reject unauthorized connections.
+    // Does not apply to Cloudflare Workers.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.rejectUnauthorized');
+  }
+
+  if (tlsOptions.ALPNProtocols) {
+    // Does not apply to Cloudflare Workers.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.ALPNProtocols');
+  }
+
+  if (tlsOptions.SNICallback) {
+    // Does not apply to Cloudflare Workers.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.SNICallback');
+  }
+
+  if (tlsOptions.requestOCSP) {
+    // Not yet supported. Can be implemented in the future.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.requestOCSP');
+  }
+
+  if (tlsOptions.secureContext) {
+    // TODO(soon): Investigate supporting this.
+    throw new ERR_OPTION_NOT_IMPLEMENTED('options.secureContext');
+  }
+
+  if (tlsOptions.pskCallback) {
```

Comment:
I would make these `if ({option} !== undefined` so that no value is accepted.

Reply:
This can be too much restrictive since some use default values (such as false). Even in tls_wrap, Node.js uses default values and passes false. For example, this will fail on `rejectUnauthorized: false` which is used a lot afaict.

Reply:
Yes, but things like { pskCallback: false} does not make any sense. Nor does tlsOptions.ALPNProtocols = ''.. etc. The options whose values are not booleans should not just ignore any falsy value.

Reply:
Yes. It definitely makes sense to be restrictive on them.